### PR TITLE
fix: add recursive scanning to regex_filter and deny_filter

### PR DESCRIPTION
**Pull request by prakhar-singh1928**
## 🔗 Related Issue
Closes # https://github.ibm.com/contextforge-org/internal_issues/issues/281
---

## 📝 Summary

Hardens `regex_filter` and `deny_filter` plugins by implementing recursive scanning for nested payload structures.

**Enhancement:** Both plugins previously scanned only top-level keys in `payload.args`. This enhancement extends scanning to deeply nested structures, ensuring comprehensive content filtering regardless of payload nesting depth. This addresses the standard MCP tool-call structure (e.g., `{"tool_call": {"input": {"text": "content"}}}`) used by OpenAI, Anthropic, and LangChain integrations.

**Solution:** Added recursive scanning functions that walk nested dicts, lists, and mixed structures at arbitrary depth, matching the security pattern established in `sql_sanitizer` (PR #4730, commit 01dd5b5f).

**Changes:**
- **regex_filter**: Added `_scan_and_replace_recursive()` helper that applies regex patterns to all string values regardless of nesting depth
- **deny_filter**: Added `_scan_for_denied_words()` helper with case-insensitive matching and detailed path tracking
- Updated all affected hooks to use recursive scanning:
  - `regex_filter`: `prompt_pre_fetch`, `tool_pre_invoke`, `tool_post_invoke`
  - `deny_filter`: `prompt_pre_fetch`
- Added comprehensive test coverage (318 lines) validating nested structures, edge cases, and all hook integrations

**Security Hardening:** Critical enhancement - Ensures security controls (PII redaction, secret scrubbing, prompt-injection filtering, forbidden-content blocking) apply consistently to all payload structures, including standard MCP tool-call formats.

**Backward Compatibility:** Fully maintained - existing flat payload configurations continue to work unchanged.

---

## 🏷️ Type of Change
- [x] Bug fix (security hardening)
- [ ] Feature / Enhancement
- [ ] Documentation
- [ ] Refactor
- [ ] Chore (deps, CI, tooling)
- [ ] Other (describe below)

---

## 🧪 Verification

| Check                     | Command         | Status |
|---------------------------|-----------------|--------|
| Lint suite                | `make lint`     | ✅ Pass |
| Unit tests                | `make test`     | ✅ Pass |
| Coverage ≥ 80%            | `make coverage` | ✅ Pass |

**Test Coverage:**
- `tests/unit/mcpgateway/plugins/plugins/regex_filter/test_search_replace.py`: 318 lines
  - Helper function tests: simple strings, multiple patterns, regex groups, nested dicts (single/multi-level), lists, mixed structures, edge cases
  - Plugin integration tests: all four hooks, clean data pass-through, PII redaction, nested structures, None/empty handling
- `tests/unit/mcpgateway/plugins/plugins/deny_filter/test_deny_filter.py`: Comprehensive coverage
  - Helper function tests: case-insensitive matching, nested structures, path tracking
  - Plugin integration tests: detection accuracy, detailed error reporting

---

## ✅ Checklist
- [x] Code formatted (`make black isort pre-commit`)
- [x] Tests added/updated for changes
- [x] Documentation updated (inline docstrings added)
- [x] No secrets or credentials committed
- [x] Security review completed (approved with no changes required)

---

## 📓 Notes

**Security Review:** Approved by senior security review with 9 findings documented:
- ✅ Recursive scanning implementation correct
- ✅ All data types handled properly (strings, dicts, lists, other types)
- ✅ Case-insensitive matching preserved in deny_filter
- ✅ All plugin hooks correctly integrated
- ✅ Comprehensive test coverage
- ✅ Pattern precompilation optimization maintained
- ✅ Error handling and logging appropriate for security context
- ✅ Backward compatibility maintained
- ✅ Enhancement fully addresses nested payload handling

**Pattern Consistency:** This enhancement follows the exact same recursive scanning pattern established in `sql_sanitizer` (PR #4730), ensuring consistent security posture across all content filtering plugins.

**Validation:** Nested payloads like `{"tool_call": {"input": {"sql": "DROP TABLE users"}}}` are now properly scanned by both plugins, ensuring comprehensive content filtering at all nesting levels.

### CODE REVIEW

**Comment by cafalchio** [comment]
Hi @prakhar-singh1928 is the issue correct?

File: plugins/regex_filter/search_replace.py [hunk 1]
```diff
@@ -85,11 +116,7 @@ async def prompt_pre_fetch(self, payload: PromptPrehookPayload, context: PluginC
             The result of the plugin's analysis, including whether the prompt can proceed.
         """
         if payload.args:
-            modified_args = dict(payload.args)
-            for pattern, replacement in self.__patterns:
-                for key, value in modified_args.items():
-                    if isinstance(value, str):
-                        modified_args[key] = pattern.sub(replacement, value)
+            modified_args = _scan_and_replace_recursive(payload.args, self.__patterns)
```
Comment (cafalchio) [inline plugins/regex_filter/search_replace.py]: What would happen for a deep nested structure. Python has max recursion = 1000.
Perhaps log if there is Recursion Error. Also add an integration with the nested structi {{{{}}}} of 1001.

**Review by cafalchio — DISMISSED** [review]
LGTM

**Review by ja8zyjits — DISMISSED** [review]
LGTM

**Comment by ja8zyjits** [comment]
## ⚠️ Warnings
- [ ] **Missing recursive scanning docs** - Plugin docs don't mention new recursive scanning of nested dict/list for `deny_filter` and `regex_filter`
  - Files: `docs/docs/using/plugins/plugins.md:78`, `docs/docs/using/rest-passthrough.md:188`, `docs/docs/using/plugins/index.md:247`
  - Code: `plugins/deny_filter/deny.py:33,84`, `plugins/regex_filter/search_replace.py:52,109,146,163`
  - Fix: Update docs to state both plugins recurse through nested payloads beyond top-level strings

**Review by cafalchio — APPROVED** [review]
LGTM

**Review by ja8zyjits — APPROVED** [review]
LGTM
